### PR TITLE
[ESIMD] Add compile-time parameter checks to lsc APIs.

### PR DESCRIPTION
**Pull request by kbobrovs**
Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 1]
```diff
@@ -915,6 +915,52 @@ __ESIMD_API void lsc_block_store(T *p, __ESIMD_NS::simd<T, NElts> vals) {
...
+  static_assert(IsStore || GRFByteSize <= 2048,
+                "2D load supports 2048 bytes max");
+  static_assert(!Transposed || !Transformed,
+                "Transposed and transformed is not supported");
+  static_assert(!Transposed || (Transposed && NBlocks == 1),
+                "Transposed expected to be 1 block only");
+  static_assert(!Transformed || (sizeof(T) == 1 || sizeof(T) == 2),
+                "VNNI transform is supported only for data size u8 or u16");
+  static_assert(!Transposed || (sizeof(T) == 4 || sizeof(T) == 8),
+                "Transposed load is supported only for data size u32 or u64");
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:935]: Just wanted to check. Transposed load handles 4 and 8 byte elements. Is transposed _store_ works for all data sizes?

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 2]
```diff
@@ -915,6 +915,52 @@ __ESIMD_API void lsc_block_store(T *p, __ESIMD_NS::simd<T, NElts> vals) {
...
+          bool Transposed, bool Transformed, bool IsStore = false>
+constexpr void check_lsc_block_2d_restrictions() {
+  constexpr int GRFByteSize = BlockWidth * BlockHeight * NBlocks * sizeof(T);
+  static_assert(!IsStore || GRFByteSize <= 512,
+                "2D store supports 512 bytes max");
+  static_assert(IsStore || GRFByteSize <= 2048,
+                "2D load supports 2048 bytes max");
+  static_assert(!Transposed || !Transformed,
+                "Transposed and transformed is not supported");
+  static_assert(!Transposed || (Transposed && NBlocks == 1),
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:930]: This line is equivalent to the shorter version:
```suggestion
  static_assert(!Transposed || NBlocks == 1,
```
The checks verifying _only one_ of {Transposed,Transformed} would be more readable being placed under those 'if constexpr (Transposed)' below at L936 and 'if constexpr (Transformed)' at L943

**Review by v-klochkov — APPROVED** [review]

**Comment by v-klochkov** [comment]
It needs clang-format fix

**Comment by kbobrovs** [comment]
The single failure on SYCL / Linux / HIP AMDGPU  is unrelated:
Failed Tests (1):
  SYCL :: Reduction/reduction_span.cpp

**Comment by kbobrovs** [comment]
@v-klochkov, please approve

**Review by v-klochkov — APPROVED** [review]

**Pull request merged**